src/MetadataManager/EventManager.ts

```diff
@@ -0,0 +1,300 @@
+
+
+class EventQueue {
+  private inProgress: boolean = false;
+  private queue: {method: any, input: {[key: string]: any}}[] = [];
+
+  constructor() {
+    this.queue = [];
+  }
+
+  enqueue(method: any, input: {[key: string]: any}): void {
+    this.queue.push({method: method, input: input});
+    this.autoAction();
+  }
+
+  front() {
+    return this.queue[0];
+  }
+  dequeue() {
+    this.queue.shift();
+  }
+
+  clear() {
+    this.inProgress = false;
+    this.queue = [];
+  }
+
+  isEmpty() {
+    return this.queue.length === 0;
+  }
+
+  autoAction() {
+    if (this.inProgress === false) {
+      this.inProgress = true;
+      this.action();
+    }
+  }
+
+  async action() {
+    await this.front().method();
+    this.dequeue();
+
+    if (this.isEmpty()) {
+      this.clear();
+    } else {
+      this.action();
+    }
+  }
+}
+
+class EventBuffer {
+  private _queue = new EventQueue();
+  constructor() {}
+  public setEvent(request: any, input: {[key: string]: any}) {
+    this._queue.enqueue(() => {
+      return new Promise(() => {
+        if (Object.keys(input).length === 0) {
+          request();
+        } else {
+          request();
+        }
+      });
+    }, input);
+  }
+}
+
+
+
+/* istanbul ignore next */
+export class MetadataEventManager {
+  private fileWatcher = vscode.workspace.createFileSystemWatcher(`**/*`);  // glob pattern
+  private eventBuffer = new EventBuffer();
+
+  public static didCreateUri: vscode.Uri|undefined = undefined;
+
+  public static register(context: vscode.ExtensionContext) {
+    let workspaceRoot: vscode.Uri|undefined = undefined;
+
+    try {
+      workspaceRoot = vscode.Uri.file(obtainWorkspaceRoot());
+      Logger.info('OneExplorer', `workspace: ${workspaceRoot.fsPath}`);
+    } catch (e: unknown) {
+      if (e instanceof Error) {
+        if (e.message === 'Need workspace') {
+          Logger.info('OneExplorer', e.message);
+        } else {
+          Logger.error('OneExplorer', e.message);
+          Balloon.error('Something goes wrong while setting workspace.', true);
+        }
+      } else {
+        Logger.error('OneExplorer', 'Unknown error has been thrown.');
+      }
+    }
+
+    const manager = new MetadataEventManager();
+
+    let registrations = [
+      manager.fileWatcher.onDidChange(async uri => {
+        if (workspaceRoot) {
+          manager.eventBuffer.setEvent(manager.changeFileEvent, {'uri': uri});
+        }
+      }),
+
+      manager.fileWatcher.onDidDelete(async uri => {
+        const pathToHash = await PathToHash.getInstance();
+        if (!pathToHash.exists(uri)) {
+          return;
+        }
+
+        const toUri = MetadataEventManager.didCreateUri;
+        if (toUri) {
+          MetadataEventManager.didCreateUri = undefined;
+          // The file/folder is moved/renamed
+          // case 1. [Dir]+Path       | move > search (delete & new)
+          if (fs.statSync(toUri.path).isDirectory()) {
+            manager.eventBuffer.setEvent(
```

Comment:
It's so cool that you have implemented this eventBuffer, and the code is far readable and safe. I am so proud :rocket: